pkg/server/home_workspaces.go

```diff
@@ -0,0 +1,656 @@
+		// If we're checking a home workspace or home bucket workspace, but not of the consistent user, let's refuse.
+		klog.Errorf("failed to authorize user %q to create a home workspace %q: home workspace can only be created by the user of the home workspace", user.GetName(), lcluster.Name)
+		responsewriters.Forbidden(ctx, attributes, rw, req, authorization.WorkspaceAcccessNotPermittedReason, homeWorkspaceCodecs)
+		return
+	}
+
+	// Test if the user has the right to create his Home workspace when it doesn't exists
+	// => test the create verb on the clusterworkspaces/workspace subresource named ~ in the root workspace.
+
+	if decision, reason, err := h.authz.Authorize(
+		request.WithCluster(ctx, request.Cluster{Name: tenancyv1alpha1.RootCluster}),
+		attributes); err != nil {
+		klog.Errorf("failed to authorize user %q to create a home workspace %q: %w", user.GetName(), lcluster.Name, err)
+		responsewriters.Forbidden(ctx, attributes, rw, req, authorization.WorkspaceAcccessNotPermittedReason, homeWorkspaceCodecs)
+		return
+	} else if decision != authorizer.DecisionAllow {
+		responsewriters.Forbidden(ctx, attributes, rw, req, reason, homeWorkspaceCodecs)
+		return
+	}
+
+	if retryAfterSeconds, err := h.tryToCreate(ctx, user.GetName(), lcluster.Name, workspaceType); err != nil {
+		responsewriters.ErrorNegotiated(err, errorCodecs, schema.GroupVersion{}, rw, req)
+		return
+	} else {
+		// Return a 429 status asking the client to try again after the creationDelay
+		rw.Header().Set("Retry-After", fmt.Sprintf("%d", retryAfterSeconds))
+		http.Error(rw, "Creating the home workspace", http.StatusTooManyRequests)
+		return
+	}
+}
+
+// reHomeWorkspaceNameDisallowedChars is the regexp that defines what characters
+// are disallowed in a home workspace name.
+// Home workspace name is derived from the user name, with disallowed characters
+// replaced.
+var reHomeWorkspaceNameDisallowedChars = regexp.MustCompile("[^a-z0-9-]")
+
+// getHomeLogicalClusterName returns the logicalcluster name of the home workspace for a given user
+// The home workspace logical cluster ancestors are home bucket workspaces whose name is based
+// on the user name sha1 hash.
+func (h *homeWorkspaceHandler) getHomeLogicalClusterName(userName string) logicalcluster.Name {
+	// bucketLevels <= 5
+	// bucketSize <= 4
+	bytes := sha1.Sum([]byte(userName))
+
+	result := h.homePrefix
+	for level := 0; level < h.bucketLevels; level++ {
+		var bucketBytes = make([]byte, h.bucketSize)
+		bucketBytesStart := level
+		bucketCharInteger := binary.BigEndian.Uint32(bytes[bucketBytesStart : bucketBytesStart+4])
+		for bucketCharIndex := 0; bucketCharIndex < h.bucketSize; bucketCharIndex++ {
+			bucketChar := byte('a') + byte(bucketCharInteger%26)
+			bucketBytes[bucketCharIndex] = bucketChar
+			bucketCharInteger /= 26
+		}
+		result = result.Join(string(bucketBytes))
+	}
+
+	userName = reHomeWorkspaceNameDisallowedChars.ReplaceAllLiteralString(userName, "-")
```

Comment:
this risks name conflicts of users.1
1. do we handle conflict securely? It is maybe ok to accept conflict and 0.0001% of users get broken. But it would not be ok to give two users access to the same workspace. Storing the full username on the workspace object as annotation could help.
2. any sensible way to deconflict? Is it worth it?

Reply:
done 1